opengrok-indexer/src/main/java/org/opengrok/indexer/history/FileHistoryCache.java

```diff
@@ -103,38 +103,34 @@ private void doFileHistory(String filename, List<HistoryEntry> historyEntries,
             Repository repository, File srcFile, File root, boolean renamed)
             throws HistoryException {
 
-        History hist = null;
+        File file = new File(root, filename);
+        // Only store directory history for the top-level directory.
+        if (file.isDirectory() && !filename.equals(repository.getDirectoryName())) {
+            LOGGER.log(Level.FINE, "Not storing history cache for {0}", file);
+            return;
+        }
 
+        History hist = null;
         /*
          * If the file was renamed (in the changesets that are being indexed),
          * its history is not stored in the historyEntries so it needs to be acquired
          * directly from the repository.
-         * This ensures that complete history of the file (across renames)
-         * will be saved.
+         * This ensures that complete history of the file (across renames) will be saved.
          */
         if (renamed) {
             hist = repository.getHistory(srcFile);
```

Comment:
What is the contract here? Can `repository.getHistory(srcFile)` return `null`? If not, then maybe we can rewrite it to:
History hist;
```java
if (renamed) {
    hist = repository.getHistory(srcFile);
} else {
    hist = new History(historyEntries);
}
```

Reply:
I thought about the `null` return value too. It seems rather than returning null History the underlying methods will throw `HistoryException`.